### PR TITLE
Three step CWLs in PAPI

### PR DESCRIPTION
**Pull request by mcovarr**
This runs locally but fails in Travis with no diagnostics and I don't know why yet. I'm assuming that's an issue with the Travis environment that's orthogonal to what's in these changes now. Inputs are still broken so there's still work to do before useful CWLs can run on PAPI.

### CODE REVIEW

File: backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala [hunk 1]
```diff
@@ -215,8 +216,11 @@ trait StandardAsyncExecutionActor extends AsyncBackendJobExecutionActor with Sta
         |(
         |cd $cwd
         |INSTANTIATED_COMMAND
-        |)
+        |) > $stdoutPath 2> $stderrPath
```
Comment (mcovarr) [inline backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala]: need to capture to controlled filenames in the main script, not just the wrapper script that's only for shared filesystems.

File: backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala [hunk 2]
```diff
@@ -248,7 +252,7 @@ trait StandardAsyncExecutionActor extends AsyncBackendJobExecutionActor with Sta
   def redirectOutputs(command: String): String = {
     // > 128 is the cutoff for signal-induced process deaths such as might be observed with abort.
     // http://www.tldp.org/LDP/abs/html/exitcodes.html
-    s"""$command > ${jobPaths.stdout} 2> ${jobPaths.stderr} < /dev/null || { rc=$$?; if [ "$$rc" -gt "128" ]; then echo $$rc; else echo -1; fi } > ${jobPaths.returnCode}"""
+    s"""$command < /dev/null || { rc=$$?; if [ "$$rc" -gt "128" ]; then echo $$rc; else echo -1; fi } > ${jobPaths.returnCode}"""
```
Comment (mcovarr) [inline backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala]: remove outer script stdout and stderr redirection

File: backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala [hunk 3]
```diff
@@ -215,8 +216,11 @@ trait StandardAsyncExecutionActor extends AsyncBackendJobExecutionActor with Sta
         |(
         |cd $cwd
         |INSTANTIATED_COMMAND
-        |)
+        |) > $stdoutPath 2> $stderrPath
         |echo $$? > $rcTmpPath
+        |# Still need to dump to stdout and stderr for PAPI
+        |cat $stdoutPath
+        |>&2 cat $stderrPath
```
Comment (mcovarr) [inline backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala]: yeah ugh

File: cwl/src/main/scala/cwl/CommandOutputBinding.scala [hunk 4]
```diff
@@ -39,9 +39,10 @@ case class CommandOutputBinding(
     val womMapType = WomMapType(WomStringType, WomStringType)
     val womMaps = paths map { path =>
       // TODO: WOM: basename/dirname/size/checksum/etc.
+      val globPathWithDirectory = GlobFunctions.prefixWithGlobDir(path)
 
       val contents: Map[WomValue, WomValue] =
-        if (loadContents) Map(WomString("contents") -> WomString(load64KiB(path, ioFunctionSet))) else Map.empty
+        if (loadContents) Map(WomString("contents") -> WomString(load64KiB(globPathWithDirectory, ioFunctionSet))) else Map.empty
```
Comment (mcovarr) [inline cwl/src/main/scala/cwl/CommandOutputBinding.scala]: Make sure to prefix the filename with the glob directory. This was the "accidental success" on the shared filesystem where this code was picking up the version of the stdout file that it shouldn't have seen.

File: cwl/src/main/scala/cwl/CwlExpressionCommandPart.scala [hunk 5]
```diff
@@ -40,13 +41,13 @@ case class CommandLineBindingCommandPart(argument: CommandLineBinding) extends C
...
       val pc = ParameterContext(runtime = runtimeEnvironment.cwlMap).withInputs(inputsMap.map({
-        case (LocalName(localName), WomSingleFile(path)) => localName -> WomString(path)
+        case (LocalName(localName), sf: WomSingleFile) => localName -> valueMapper(sf)
         case (LocalName(localName), value) => localName -> value
       }), functions)
 
       val womValue: WomValue = argument match {
         case CommandLineBinding(_, _, _, _, _, Some(StringOrExpression.Expression(expression)), Some(false)) =>
-          expression.fold(EvaluateExpression).apply(pc)
+          expression.fold(EvaluateExpression).apply(pc) |> valueMapper
```
Comment (mcovarr) [inline cwl/src/main/scala/cwl/CwlExpressionCommandPart.scala]: some more things needed value mapping

File: backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala (hunk 1 above)
Comment (cjllanwarne) [reply backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala]: Mea culpa!

File: backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala (hunk 3 above)
Comment (cjllanwarne) [reply backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala]: Hmm, at the back of my memory is that PAPI might have a special field for specifying `stdout` and `stderr`, and we use that to allow the auto-magical "real-time-stdout"ing

File: cwl/src/main/scala/cwl/CommandLineTool.scala [hunk 6]
```diff
@@ -85,7 +86,6 @@ case class CommandLineTool private(
         pull <- requirement.dockerPull.orElse(requirement.dockerImageId)
       } yield ValueAsAnExpression(WomString(pull))
 
-      import mouse.option._
```
Comment (cjllanwarne) [inline cwl/src/main/scala/cwl/CommandLineTool.scala]: ToL: curious why you wanted to make it less well contained?

File: cwl/src/main/scala/cwl/CwlExpressionCommandPart.scala (hunk 5 above)
Comment (cjllanwarne) [reply cwl/src/main/scala/cwl/CwlExpressionCommandPart.scala]: Sigh... I thought the whole point of `cats` was that they did away with all these nonsense operators

File: cwl/src/main/scala/cwl/CwlExpressionCommandPart.scala [hunk 7]
```diff
@@ -66,10 +67,8 @@ case class InputParameterCommandPart(commandInputParameter: CommandInputParamete
       val womValue: WomValue = commandInputParameter match {
         case cip: CommandInputParameter =>
           val localizedId = LocalName(FullyQualifiedName(cip.id).id)
-          inputsMap.get(localizedId) match {
-            case Some(x) => valueMapper(x)
-            case _ => throw new RuntimeException(s"could not find $localizedId in map $inputsMap")
-          }
+          inputsMap.get(localizedId).cata(
```
Comment (cjllanwarne) [inline cwl/src/main/scala/cwl/CwlExpressionCommandPart.scala]: ... and what's a `cata`?

File: backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala (hunk 3 above)
Comment (mcovarr) [reply backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala]: Why do I have this sinking feeling that you're right. 😦 I may be able to get around that with `tee` but I was hoping not to have to introduce that dependency. I'll look into this more.

File: cwl/src/main/scala/cwl/CommandLineTool.scala (hunk 6 above)
Comment (mcovarr) [reply cwl/src/main/scala/cwl/CommandLineTool.scala]: Yeah I might move it back. I think there was an intermediate step where I was using this in more than one spot, but that's clearly not the case anymore.

File: cwl/src/main/scala/cwl/CwlExpressionCommandPart.scala (hunk 5 above)
Comment (mcovarr) [reply cwl/src/main/scala/cwl/CwlExpressionCommandPart.scala]: Personally I have learned to love the right turn arrow's effect on readability. 😄

File: cwl/src/main/scala/cwl/CwlExpressionCommandPart.scala (hunk 7 above)
Comment (mcovarr) [reply cwl/src/main/scala/cwl/CwlExpressionCommandPart.scala]: That's how its close friends refer to a [catamorphism](https://en.wikipedia.org/wiki/Catamorphism).

File: cwl/src/main/scala/cwl/CwlExpressionCommandPart.scala (hunk 5 above)
Comment (cjllanwarne) [reply cwl/src/main/scala/cwl/CwlExpressionCommandPart.scala]: After a short stay in Room `1|>1`?

**Pull request merged**